core/coreunix/add.go

```diff
@@ -153,112 +153,148 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
 }
 
 func (adder *Adder) RootNode() (node.Node, error) {
```

Comment:
Does this make sense as well? It feels cleaner, it compiles and the tests pass.
```diff
diff --git a/core/coreunix/add.go b/core/coreunix/add.go
index 516b76d44..090ecee27 100644
--- a/core/coreunix/add.go
+++ b/core/coreunix/add.go
@@ -152,40 +152,42 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
        return balanced.BalancedLayout(params.New(chnk))
 }
 
-func (adder *Adder) RootNode() (node.Node, error) {
+func (adder *Adder) RootNode() (root node.Node, err error) {
        eip := log.EventBegin(adder.ctx, "RootNode")
-       defer eip.Done()
+       defer func() {
+               if root != nil {
+                       eip.Append(root.Cid())
+               }
+               if err != nil {
+                       eip.SetError(err)
+               }
+               eip.Done()
+       }()
        // for memoizing
-       if adder.root != nil {
-               eip.Append(adder.root.Cid())
-               return adder.root, nil
+       root = adder.root
+       if root != nil {
+               return root, nil
        }
 
        mr, err := adder.mfsRoot()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
-       root, err := mr.GetValue().GetNode()
+       root, err = mr.GetValue().GetNode()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
 
        // if not wrapping, AND one root file, use that hash as root.
        if !adder.Wrap && len(root.Links()) == 1 {
-               nd, err := root.Links()[0].GetNode(adder.ctx, adder.dagService)
+               root, err = root.Links()[0].GetNode(adder.ctx, adder.dagService)
                if err != nil {
-                       eip.SetError(err)
                        return nil, err
                }
-
-               root = nd
        }
 
        adder.root = root
-       eip.Append(root.Cid())
-       return root, err
+       return
 }
```

Reply:
I'm fine with naming the return values for the purpose of accessing them from within the `defer` statement. However, I wouldn't use the implicit return feature (IMO, it was a mistake). How about:

```
diff --git a/core/coreunix/add.go b/core/coreunix/add.go
index 516b76d44..090ecee27 100644
--- a/core/coreunix/add.go
+++ b/core/coreunix/add.go
@@ -152,37 +152,38 @@ func (adder Adder) add(reader io.Reader) (node.Node, error) {
        return balanced.BalancedLayout(params.New(chnk))
 }
 
-func (adder *Adder) RootNode() (node.Node, error) {
+func (adder *Adder) RootNode() (root node.Node, err error) {
        eip := log.EventBegin(adder.ctx, "RootNode")
-       defer eip.Done()
+       defer func() {
+               if root != nil {
+                       eip.Append(root.Cid())
+               }
+               if err != nil {
+                       eip.SetError(err)
+               }
+               eip.Done()
+       }()
        // for memoizing
-       if adder.root != nil {
-               eip.Append(adder.root.Cid())
-               return adder.root, nil
+       if addr.root != nil {
+               return addr.root, nil
        }
 
        mr, err := adder.mfsRoot()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
-       root, err := mr.GetValue().GetNode()
+       root, err = mr.GetValue().GetNode()
        if err != nil {
-               eip.SetError(err)
                return nil, err
        }
 
        // if not wrapping, AND one root file, use that hash as root.
        if !adder.Wrap && len(root.Links()) == 1 {
-               nd, err := root.Links()[0].GetNode(adder.ctx, adder.dagService)
+               root, err = root.Links()[0].GetNode(adder.ctx, adder.dagService)
                if err != nil {
-                       eip.SetError(err)
                        return nil, err
                }
-
-               root = nd
        }
 
        adder.root = root
-       eip.Append(root.Cid())
        return root, err
 }
```

Reply:
agreed, the implicit return thing was a terrible choice by the go team

Reply:
Alright, I can make these changes on the rest of the methods here too then. I am not sure if I should make that a separate commit or rebase. What would you prefer?

Reply:
either way, i'm not picky. Though I do appreciate smaller numbers of commits